docs/decisions/0005-architecture-and-data-modeling.rst

```diff
@@ -0,0 +1,149 @@
+
+Use the Enforcement API for Authorization Decisions for External Clients
+------------------------------------------------------------------------
+- External clients (e.g., MFEs, IDAs, or any service not co-located with the policy store) must use the REST API provided by the Open edX authorization layer to request authorization decisions.
+
+Use a Stable and Versioned Public API for Other Clients
+-------------------------------------------------------
+- The Open edX Layer will provide a stable and versioned Public API for services to interact with the authorization system. This API will be well-documented and include clear contracts for inputs, outputs, and error handling.
+- In-process clients (e.g., LMS, CMS, or any service co-located with the policy store) may use this Public API (``api.py``) directly to interact with the Open edX Layer without going through the REST API, but they must still adhere to the same contracts and versioning.
+- Clients must provide all necessary context for authorization decisions, including subject, action, object, and any relevant contextual information (e.g., organization, course) and the authorization layer will make the decision based on the policies in the policy store.
+
+Consequences
+************
+
+#. **New Components in the Open edX Ecosystem**: There are several new components introduced as part of this architecture:
+   - Policy Store: A centralized datastore for storing and managing authorization policies.
+   - Enforcement API: An API for enforcing authorization policies and making authorization decisions.
+   - Policy Management API: An API for creating, updating, and deleting dynamic policies in the Policy Store.
+   - Open edX Layer: The layer that abstracts access to the Policy Store and provides a unified interface for authorization.
+   - Authorization Engine: The Casbin-based engine that evaluates authorization requests based on defined policies.
+
+#. **Services Should be Migrated to Use this new Architecture**: Existing services that currently implement their own authorization logic will need to be migrated to use the new architecture. This may involve:
+   - Refactoring code to remove direct authorization checks and replace them with calls to the Enforcement API.
+   - Defining policies in the ``authz.policy`` file and/or via the Policy Management API.
+   - Ensuring that all necessary context is provided when making authorization requests.
+
+#. **The Framework Requires Client Integration**: To make authorization decisions clients must:
+   - Call the Open edX Layer's Enforcement API via network with a valid token for authentication and authorization.
+   - Install the necessary libraries (e.g., openedx-authz) as dependencies to interact with the Open edX Layer directly if they are in-process clients via the Public API (``api.py``). These libraries should be part of the services's ``INSTALLED_APPS`` -for data migration purposes- and requirements files.
```

Comment:
Should there be a client API in the library that can present the same / similar API to external services as what LMS/CMS get? It could wrap the REST calls, but in the end should be providing the same functionality.